File: cscorlib/CodeModels/CLI/Signature/TypeSig.cs
using System;
using System.Collections;
using CooS.Reflection;
using CooS.CodeModels.CLI.Metadata;

namespace CooS.CodeModels.CLI.Signature {

	class TypeSig {

		ElementType elementType;
		TypeDefOrRefEncoded typeDefOrRef;
		CustomMod[] customMods;
		TypeSig type;
		bool _void;
		bool _byref;
		bool _pinned;
		bool _typedbyref;
		MethodSig method;

		public TypeSig(SignatureReader reader) {
			//COOS: CLI仕様にはないんだけど、FreeType2のアセンブリに出現したので。
			customMods = SignatureUtility.ReadCustomMods(reader);
			//END OF COSTOMIZE
			elementType = reader.ReadMark();
			switch(elementType) {
			default:
				reader.Dump();
				throw new BadSignatureException("Invalid ElementType: "+(int)elementType);
			case ElementType.Boolean:
			case ElementType.Char:
			case ElementType.I1:
			case ElementType.U1:
			case ElementType.I2:
			case ElementType.U2:
			case ElementType.I4:
			case ElementType.U4:
			case ElementType.I8:
			case ElementType.U8:
			case ElementType.R4:
			case ElementType.R8:
			case ElementType.I:
			case ElementType.U:
			case ElementType.String:
			case ElementType.Object:
				return;
			case ElementType.ValueType:
			case ElementType.Class:
				//| VALUETYPE TypeDefOrRefEncoded
				//| CLASS TypeDefOrRefEncoded
				typeDefOrRef = new TypeDefOrRefEncoded(reader);
				return;
			case ElementType.Ptr:
				//| PTR CustomMod* VOID
				//| PTR CustomMod* TypeSig
				customMods = SignatureUtility.ReadCustomMods(reader);
				type = new TypeSig(reader);
				return;
			case ElementType.FnPtr:
				//| FNPTR MethodDefSig
				//| FNPTR MethodRefSig
				this.method = new MethodSigImpl(reader);
				return;
			case ElementType.Array:
				//| ARRAY TypeSig ArrayShape  (general array, see clause 22.2.13)
				throw new NotImplementedException("Not Implemented: ElementType.ARRAY");
			case ElementType.SzArray:
				//| SZARRAY CustomMod* TypeSig (single dimensional, zero-based array i.e. vector)
				customMods = SignatureUtility.ReadCustomMods(reader);
				type = new TypeSig(reader);
				return;
			case ElementType.ByRef:			//COOS: Is this correct?
				_byref = true;
				type = new TypeSig(reader);
				return;
			case ElementType.Void:			//COOS: Is this correct?
				_void = true;
				return;
			case ElementType.TypedByRef:	//COOS: Is this correct?
				_typedbyref = true;
				return;
			case ElementType.Pinned:		//COOS: Is this correct?
				_pinned = true;
				type = new TypeSig(reader);
				return;
			case ElementType.CModReqd:
			case ElementType.CModOpt:
			case ElementType.Internal:
			case ElementType.Modifier:
			case ElementType.Sentinel:
				reader.Dump();
				throw new BadSignatureException("TypeSig::Parse detects unsupported beginning: 0x"+((int)elementType).ToString("X"));
			}
		}

		public override string ToString() {
			return "TypeSig:0x"+((int)this.elementType).ToString("X2")+","+this.typeDefOrRef;
		}

		public ElementType ElementType {
			get {
				return this.elementType;
			}
		}

		public CustomMod[] CustomMods {
			get {
				return this.customMods;
			}
		}

		public TypeSig Type {
			get {
				return this.type;
			}
		}

		public TypeDefOrRefEncoded TypeDefOrRef {
			get {
				return this.typeDefOrRef;
			}
		}

		public MethodSig Method {
			get {
				return this.method;
			}
		}

		public bool Void {
			get {
				return this._void;
			}
		}

		public bool ByRef {
			get {
				return this._byref;
			}
		}

		public bool TypedByRef {
			get {
				return this._typedbyref;
			}
		}

		public TypeImpl ResolveTypeAt(AssemblyDef assembly) {
			if(assembly==null) throw new ArgumentNullException("assembly");
			switch(elementType) {
			default:
				throw new BadSignatureException();
			case ElementType.Boolean:
				return assembly.Manager.Boolean;
			case ElementType.Char:
				return assembly.Manager.Char;
			case ElementType.I1:
				return assembly.Manager.SByte;
			case ElementType.U1:
				return assembly.Manager.Byte;
			case ElementType.I2:
				return assembly.Manager.Int16;
			case ElementType.U2:
				return assembly.Manager.UInt16;
			case ElementType.I4:
				return assembly.Manager.Int32;
			case ElementType.U4:
				return assembly.Manager.UInt32;
			case ElementType.I8:
				return assembly.Manager.Int64;
			case ElementType.U8:
				return assembly.Manager.UInt64;
			case ElementType.R4:
				return assembly.Manager.Single;
			case ElementType.R8:
				return assembly.Manager.Double;
			case ElementType.I:
				return assembly.Manager.IntPtr;
			case ElementType.U:
				return assembly.Manager.UIntPtr;
			case ElementType.String:
				return assembly.Manager.String;
			case ElementType.Object:
				return assembly.Manager.Object;
			case ElementType.Void:
				return assembly.Manager.Void;
			case ElementType.TypedByRef:
				return assembly.Manager.ResolveType("System.TypedReference",true);
			case ElementType.ValueType:
			case ElementType.Class:
				return this.typeDefOrRef.ResolveTypeAt(assembly);
			case ElementType.ByRef:
				return (SuperType)type.ResolveTypeAt(assembly).GetByRefPointerType();
			case ElementType.Ptr:
				return (SuperType)type.ResolveTypeAt(assembly).GetByValPointerType();
			case ElementType.SzArray:
				return (SuperType)type.ResolveTypeAt(assembly).GetSzArrayType();
			case ElementType.Array:
				throw new NotImplementedException("Not Implemented: ElementType.ARRAY");
			case ElementType.FnPtr:
				throw new NotImplementedException("Not Implemented: ElementType.FNPTR");
			case ElementType.Pinned:
				return this.type.ResolveTypeAt(assembly);
			case ElementType.CModReqd:
			case ElementType.CModOpt:
			case ElementType.Internal:
			case ElementType.Modifier:
			case ElementType.Sentinel:
				throw new BadSignatureException("TypeSig::Parse detects unsupported beginning: "+(int)elementType);
			}
		}

	}

}

File: kernel/signature/TypeSig.cpp
#include "TypeSig.h"
#include "TypeDefOrRefEncoded.h"
#include "MethodDefSig.h"
#include "MethodRefSig.h"
#include "ArrayShape.h"


namespace Signature {

	TypeSig::TypeSig() {
		TypeDefOrRef = NULL;
		MethodDefSig = NULL;
		MethodRefSig = NULL;
		Type = NULL;
		ArrayShape = NULL;
		Void = false;
	}

	TypeSig::~TypeSig() {
		delete TypeDefOrRef;
		delete MethodDefSig;
		delete MethodRefSig;
		delete Type;
		delete ArrayShape;
	}

	TypeSig::TypeSig(const TypeSig& t) : CustomMods(t.CustomMods) {
		ElementType = t.ElementType;
		Void = t.Void;
		TypeDefOrRef = t.TypeDefOrRef ? new TypeDefOrRefEncoded(*t.TypeDefOrRef) : NULL;
		MethodDefSig = t.MethodDefSig ? new Signature::MethodDefSig(*t.MethodDefSig) : NULL;
		MethodRefSig = t.MethodRefSig ? new Signature::MethodRefSig(*t.MethodRefSig) : NULL;
		Type = t.Type ? new TypeSig(*t.Type) : NULL;
		ArrayShape = t.ArrayShape ? new Signature::ArrayShape(*t.ArrayShape) : NULL;
	}

	bool TypeSig::Parse(SignatureStream& _ss) {
		SignatureStream ss(_ss);
		ElementType = (ELEMENT_TYPE)ss.readByte();
		switch(ElementType) {
		default:
			return false;
		case ELEMENT_TYPE_BOOLEAN:
		case ELEMENT_TYPE_CHAR:
		case ELEMENT_TYPE_I1:
		case ELEMENT_TYPE_U1:
		case ELEMENT_TYPE_I2:
		case ELEMENT_TYPE_U2:
		case ELEMENT_TYPE_I4:
		case ELEMENT_TYPE_U4:
		case ELEMENT_TYPE_I8:
		case ELEMENT_TYPE_U8:
		case ELEMENT_TYPE_R4:
		case ELEMENT_TYPE_R8:
		case ELEMENT_TYPE_I:
		case ELEMENT_TYPE_U:
		case ELEMENT_TYPE_STRING:
		case ELEMENT_TYPE_OBJECT:
			ss.commit();
			return true;
		case ELEMENT_TYPE_VALUETYPE:
		case ELEMENT_TYPE_CLASS:
			//| VALUETYPE TypeDefOrRefEncoded
			//| CLASS TypeDefOrRefEncoded
			TypeDefOrRef = new TypeDefOrRefEncoded();
			if(!TypeDefOrRef->Parse(ss)) {
				return false;
			} else {
				ss.commit();
				return true;
			}
		case ELEMENT_TYPE_BYREF:	// Is this correct?
		case ELEMENT_TYPE_PTR:
			//| PTR CustomMod* VOID
			//| PTR CustomMod* TypeSig
			while(true) {
				CustomMod cm;
				if(!cm.Parse(ss)) {
					break;
				}
				CustomMods.push_back(cm);
			}
			if(ss.peekByte()==ELEMENT_TYPE_VOID) {
				Void = true;
				ss.readByte();
				ss.commit();
				return true;
			} else {
				Void = false;
				Type = new TypeSig();
				if(Type->Parse(ss)) {
					ss.commit();
					return true;
				} else {
					return false;
				}
			}
		case ELEMENT_TYPE_FNPTR:
			//| FNPTR MethodDefSig
			//| FNPTR MethodRefSig
			MethodDefSig = new Signature::MethodDefSig();
			MethodRefSig = new Signature::MethodRefSig();
			if(MethodDefSig->Parse(ss)) {
				delete MethodRefSig;
				MethodRefSig = NULL;
				ss.commit();
				return true;
			} else if(MethodRefSig->Parse(ss)) {
				delete MethodDefSig;
				MethodDefSig = NULL;
				ss.commit();
				return true;
			} else {
				delete MethodRefSig;
				delete MethodDefSig;
				MethodRefSig = NULL;
				MethodDefSig = NULL;
				return false;
			}
		case ELEMENT_TYPE_ARRAY:
			//| ARRAY TypeSig ArrayShape  (general array, see clause 22.2.13)
			panic(L"Not Implemented: ELEMENT_TYPE_ARRAY");
		case ELEMENT_TYPE_SZARRAY:
			//| SZARRAY CustomMod* TypeSig (single dimensional, zero-based array i.e. vector)
			while(true) {
				CustomMod cm;
				if(!cm.Parse(ss)) {
					break;
				}
				CustomMods.push_back(cm);
			}
			Type = new TypeSig();
			if(Type->Parse(ss)) {
				ss.commit();
				return true;
			} else {
				return false;
			}
		case ELEMENT_TYPE_VOID:
		case ELEMENT_TYPE_PINNED:
		case ELEMENT_TYPE_TYPEDBYREF:
		case ELEMENT_TYPE_CMOD_REQD:
		case ELEMENT_TYPE_CMOD_OPT:
		case ELEMENT_TYPE_INTERNAL:
		case ELEMENT_TYPE_MODIFIER:
		case ELEMENT_TYPE_SENTINEL:
			panic("TypeSig::Parse detects unsupported beginning");
		}
	}

	std::wstring TypeSig::ToString() const {
		switch(ElementType) {
		default:
			panic("Signature::TypeSig::ToString");
		case ELEMENT_TYPE_BOOLEAN:
			return L"bool";
		case ELEMENT_TYPE_CHAR:
			return L"char";
		case ELEMENT_TYPE_I1:
			return L"sbyte";
		case ELEMENT_TYPE_U1:
			return L"byte";
		case ELEMENT_TYPE_I2:
			return L"short";
		case ELEMENT_TYPE_U2:
			return L"ushort";
		case ELEMENT_TYPE_I4:
			return L"int";
		case ELEMENT_TYPE_U4:
			return L"uint";
		case ELEMENT_TYPE_I8:
			return L"long";
		case ELEMENT_TYPE_U8:
			return L"ulong";
		case ELEMENT_TYPE_R4:
			return L"float";
		case ELEMENT_TYPE_R8:
			return L"double";
		case ELEMENT_TYPE_I:
			return L"nint";
		case ELEMENT_TYPE_U:
			return L"nuint";
		case ELEMENT_TYPE_VALUETYPE:
		case ELEMENT_TYPE_CLASS:
			{
				std::wstring retval;
				switch(ElementType) {
				case ELEMENT_TYPE_VALUETYPE:
					retval = L"[ValueType]";
					break;
				case ELEMENT_TYPE_CLASS:
					retval = L"[Class]";
					break;
				}
				return retval+L" 0x"+itos<wchar_t,16>((int)TypeDefOrRef->table,2,'0')+L":"+itos<wchar_t,10>((int)TypeDefOrRef->index);
			}
		case ELEMENT_TYPE_STRING:
			//| STRING
			return L"[String]";
		case ELEMENT_TYPE_OBJECT:
			//| OBJECT
			return L"[Object]";
		case ELEMENT_TYPE_PTR:
			//| PTR CustomMod* VOID
			//| PTR CustomMod* TypeSig
			return L"[PTR]";
		case ELEMENT_TYPE_FNPTR:
			//| FNPTR MethodDefSig
			//| FNPTR MethodRefSig
			return L"[FNPTR]";
		case ELEMENT_TYPE_ARRAY:
			//| ARRAY TypeSig ArrayShape  (general array, see clause 22.2.13)
			return L"[ARRAY]";
		case ELEMENT_TYPE_SZARRAY:
			//| SZARRAY CustomMod* TypeSig (single dimensional, zero-based array i.e. vector)
			return L"[SZARRAY]";
		}
	}

	bool TypeSig::Match(const TypeSig& type) const {
		if(ElementType!=type.ElementType) return false;
		switch(ElementType) {
		default:
			panic("TypeSig::Equals detects illegal ElementType "+itos<char,16>((uint)ElementType));
		case ELEMENT_TYPE_BOOLEAN:
		case ELEMENT_TYPE_CHAR:
		case ELEMENT_TYPE_I1:
		case ELEMENT_TYPE_U1:
		case ELEMENT_TYPE_I2:
		case ELEMENT_TYPE_U2:
		case ELEMENT_TYPE_I4:
		case ELEMENT_TYPE_U4:
		case ELEMENT_TYPE_I8:
		case ELEMENT_TYPE_U8:
		case ELEMENT_TYPE_R4:
		case ELEMENT_TYPE_R8:
		case ELEMENT_TYPE_I:
		case ELEMENT_TYPE_U:
		case ELEMENT_TYPE_STRING:
		case ELEMENT_TYPE_OBJECT:
			return true;
		case ELEMENT_TYPE_VALUETYPE:
		case ELEMENT_TYPE_CLASS:
			//| VALUETYPE TypeDefOrRefEncoded
			//| CLASS TypeDefOrRefEncoded
			return TypeDefOrRef->Match(*type.TypeDefOrRef);
		case ELEMENT_TYPE_PTR:
			//| PTR CustomMod* VOID
			//| PTR CustomMod* TypeSig
			if(Void!=type.Void) {
				return false;
			}
			if(Type && type.Type) {
				return Type->Match(*type.Type);
			} else if(Void && type.Void) {
				return true;
			} else {
				return false;
			}
		case ELEMENT_TYPE_FNPTR:
			//| FNPTR MethodDefSig
			//| FNPTR MethodRefSig
			if(MethodDefSig && type.MethodDefSig) {
				return MethodDefSig->Match(*type.MethodDefSig);
			} else if(MethodRefSig && type.MethodRefSig) {
				return MethodRefSig->Match(*type.MethodRefSig);
			} else {
				return false;
			}
		case ELEMENT_TYPE_ARRAY:
			//| ARRAY TypeSig ArrayShape  (general array, see clause 22.2.13)
			panic(L"TypeSig::Equals not implemented: ELEMENT_TYPE_ARRAY");
		case ELEMENT_TYPE_SZARRAY:
			//| SZARRAY CustomMod* TypeSig (single dimensional, zero-based array i.e. vector)
			return Type->Match(*type.Type);
		}
	}

	void TypeSig::Link(const Reflection::Assembly& assembly) {
		if(TypeDefOrRef!=NULL) TypeDefOrRef->Link(assembly);
		if(Type!=NULL) Type->Link(assembly);
		if(MethodDefSig!=NULL) MethodDefSig->Link(assembly);
		if(MethodRefSig!=NULL) MethodRefSig->Link(assembly);
	}

}
